marlowe-cli/src/Language/Marlowe/CLI/Analyze.hs

```diff
@@ -0,0 +1,1032 @@
+-- | Detect an invalid role name.
+invalidRole :: P.TokenName -> Bool
+invalidRole P.TokenName{..} = P.lengthOfByteString unTokenName > 32
+
+
+-- | Check that all tokens are valid.
+checkTokens :: ContractInstance lang era  -- ^ The bundle of contract information.
+            -> A.Value                    -- ^ A report on token validity.
+checkTokens ci =
+  putJson "Tokens"
+    [
+      "Invalid tokens" .= invalidToken `S.filter` extractFromContract ci
+    ]
+
+
+-- | Check that all roles are valid.
+checkRoles :: ContractInstance lang era  -- ^ The bundle of contract information.
+           -> A.Value                    -- ^ Action to print a report on role-name validity.
+checkRoles ci =
+  let
+    roles = extractFromContract ci
+  in
+    putJson "Role names"
+      [
+        "Invalid role names" .= invalidRole `S.filter` roles
+      , "Blank role names"   .= P.adaToken `S.member` roles
+      ]
+
+
+-- | Check that the protocol limit on maximum value in a UTxO is not violated.
+checkMaximumValue :: MonadError CliError m
+                  => Api.ProtocolParameters     -- ^ The `maxValue` protocol parameter.
+                  -> Maybe [Transaction]        -- ^ The transactions to traverse.
+                  -> ContractInstance lang era  -- ^ The bundle of contract information.
+                  -> Bool                       -- ^ Whether to include worst-case example in output.
+                  -> m A.Value                  -- ^ Action to print a report on `maxValue` validity.
+checkMaximumValue Api.ProtocolParameters{protocolParamMaxValueSize=Just maxValue} maybeTransactions ci verbose =
+  let
+    (size, worst) =
+      case maybeTransactions of
+        Just transactions -> let
+                               measure tx@(Transaction _ contract _ _) = [(computeValueSize $ extractAll contract, Just tx)]
+                             in
+                               maximumBy (compare `on` fst) $ foldMap measure transactions
+        Nothing           -> (computeValueSize $ extractFromContract ci, Nothing)
+  in
+    pure
+      $ putJson "Maximum value"
+      $ [
+          "Actual"     .= size
+        , "Maximum"    .= maxValue
+        , "Unit"       .= ("byte" :: String)
+        , "Invalid"    .= (size > fromEnum maxValue)
+        , "Percentage" .= (100 * fromIntegral size / fromIntegral maxValue :: Double)
+        ]
+      <> maybe [] (pure . ("Worst case" .=)) (guard verbose >> worst)
+checkMaximumValue _ _ _ _ = throwError $ CliError "Missing `maxValue` protocol parameter."
+
+
+-- | Compute the size of a multi-asset value.
+computeValueSize :: S.Set Token  -- ^ The tokens present.
+                 -> Int          -- ^ The number of bytes on the ledger.
+computeValueSize tokens =
+  let
+    nTokens = S.size tokens
+    nPolicies = S.size $ S.map (\(Token c _) -> c) tokens
+    nNames = sum . fmap P.lengthOfByteString . toList $ S.map (\(Token _ (P.TokenName n)) -> n) tokens
+    padWords x = (x + 7) `div` 8
+  in
```

Comment:
It would be nice to have some reference to the spec or explanation of this calculation.

Reply:
Sorry, too much poinfree code here. I'll expand or document.

Reply:
I explained the algorithm and better documented in f8418144e0308964a8e0681bcb302c584156b351. Thanks!

Reply:
Thanks a lot for the comments!